### PR TITLE
Sites state: add remaining Jetpack selectors

### PR DESCRIPTION
**Pull request by Tug**
This is a follow up of https://github.com/Automattic/wp-calypso/pull/8197 which aims at adding the last remaining selectors for Jetpack sites.

| Selector | matching function/prop in `lib/site` |
| --- | --- |
| canJetpackSiteManage() | site/jetpack.js - [canManage()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L49) |
| canJetpackSiteUpdateFiles() | site/utils.js - [canUpdateFiles()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L97) |
| canJetpackSiteAutoUpdateFiles() | site/utils.js - [canAutoupdateFiles()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L124) |
| canJetpackSiteAutoUpdateCore() | site/utils.js - [canAutoupdateCore()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L136) |
| hasJetpackSiteJetpackMenus() | site/jetpack.js - [hasJetpackMenus](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L40) |
| hasJetpackSiteJetpackThemes() | site/jetpack.js - [hasJetpackThemes](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L41) |
| isJetpackSiteMainNetworkSite() | site/utils.js - [isMainNetworkSite()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L148) |
| isJetpackSiteSecondaryNetworkSite() | site/jetpack.js - [isSecondaryNetworkSite()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L77) |
| verifyJetpackModulesActive() | site/jetpack.js - [verifyModulesActive()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L91) |
| getJetpackSiteRemoteManagementURL() | site/jetpack.js - [getRemoteManagementURL()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L105) |
| hasJetpackSiteCustomDomain() | site/utils.js - [hasCustomDomain()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L183) |
| getJetpackSiteFileModDisableReasons() | site/utils.js - [getSiteFileModDisableReason()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/utils.js#L69) |
| siteHasMinimumJetpackVersion() | site/jetpack.js - [hasMinimumJetpackVersion](https://github.com/Automattic/wp-calypso/blob/master/client/lib/site/jetpack.js#L33) |
### Testing Instructions
- `git checkout update/add-jetpack-site-selectors-2`
- `npm run test-client client/state/sites/test/selectors.js`
### Reviews
- [ ] Code
- [ ] Product

### CODE REVIEW

File: client/state/sites/selectors.js [hunk 1]
```diff
@@ -550,3 +552,217 @@ export function getSitePostsPage( state, siteId ) {
...
+
+/**
+ * Returns an explanation on why updates are disabled on a Jetpack Site
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @param {String} action The update action we wanted to perform on this site
+ * @return {?String} The reason why file update is disabled
+ */
+export function getJetpackSiteFileModDisableReason( state, siteId, action = 'modifyFiles' ) {
```
Comment (Tug) [inline client/state/sites/selectors.js]: This isn't used anywhere in calypso, we will probably remove it.

**Comment by gziolo** [comment]
Circle CI complains about two failing tests. It would be nice to fix them.

Could you also provide mapping between old and new methods similar to what @retrofox done in #8197?

File: client/state/sites/test/selectors.js [hunk 2]
```diff
@@ -1184,4 +1194,675 @@ describe( 'selectors', () => {
...
+							options: {
+								file_mod_disabled: [ 'has_no_file_system_write_access' ]
+							}
+						}
+					}
+				}
+			};
+
+			const reason = getJetpackSiteFileModDisableReasons( state, siteId );
+			expect( reason ).to.deep.equal( [ i18n.translate( 'The file permissions on this host prevent editing files.' ) ] );
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: I think you can skip `translate` bits in tests. We always use english version in tests.

File: client/state/sites/test/selectors.js [hunk 3]
```diff
@@ -1184,4 +1194,675 @@ describe( 'selectors', () => {
 			expect( frontPageType ).to.eql( 'page' );
 		} );
 	} );
+
+	describe( '#canJetpackSiteManage()', () => {
+		it( 'it should return `null` for a non-existing site', () => {
+			const state = {
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: There is a bit of code duplication around initial state. What do you think about creating helper method which would create basic structure. Something like:

```js
const createState = ( items = {} ) => deepFreeze( {
    sites: {
        items
    }
} );
```

I also think it would be nice to add `deepFreeze` to make sure we never mutate state. Check other tests for reference: https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=deepFreeze.

File: client/state/sites/test/selectors.js [hunk 4]
```diff
@@ -1184,4 +1194,675 @@ describe( 'selectors', () => {
...
+			expect( canManage ).to.equal( null );
+		} );
+
+		it( 'it should return `null` for a non jetpack site', () => {
+			const siteId = 77203074;
+
+			const state = {
+				sites: {
+					items: {
+						77203074: {
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: [Object computer property](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#object-computed-properties) can be used here:
```js
[ siteId ]: { ... }
```
and other places where you use the same id.

**Comment by gziolo** [comment]
Great work so far, I'm happy to do final review once you think it's ready to be merged. As I already mention it would be nice to know what are we replacing so we could compare those functions :)

**Comment by Tug** [comment]
@gziolo Thanks for reviewing :)

> Circle CI complains about two failing tests. It would be nice to fix them.

Yes, I left this "in progress" because I'm waiting for #8197 to be merged first. Those 2 tests should be fixed after I rebase on top of it ;)

**Comment by ockham** [comment]
@Tug Any chance you merge this before #8197? I could really use one or two of your selectors for #8776, and I think this PR is in a more advanced stage than #8197.

**Comment by Tug** [comment]
@ockham Sure I could.
The conflicting part is around `canJetpackSiteUpdateFiles` which is done already in #8197 and used by another selector in this PR. I can always add the actual implementation of this function here, but it might make @retrofox remove parts of his code on rebase.

File: client/state/sites/test/selectors.js (hunk 4 above)
Comment (Tug) [inline client/state/sites/test/selectors.js]: I wouldn't try to be too clever in tests. The more static the data we give as input the better I think

**Comment by Tug** [comment]
I just cherry-picked @retrofox commits from #8197 here and addressed the comments reported there.

This should be ready for another review!

File: client/state/sites/test/selectors.js (hunk 3 above)
Comment (Tug) [inline client/state/sites/test/selectors.js]: I don't think we should be too afraid of duplicating code in tests. It indeed makes the test longer but it gives all the information of the input (and the expected output) at the same place, which is good imo.

File: client/state/sites/test/selectors.js (hunk 4 above)
Comment (gziolo) [inline client/state/sites/test/selectors.js]: This code is using const as a property value, so it doesn't change much here if you would use it also as a key to represent the same value. I agree that static data is very useful in tests, but mostly in assertions where we want to make sure that it always behaves the same.

File: client/state/sites/test/selectors.js [hunk 5]
```diff
@@ -54,20 +55,28 @@ function changeVersion( version, operator = 1 ) {
 	const splitVersion = version.split( '.' );
 
 	if ( operator >= 1 ) {
-		splitVersion[ splitVersion.length - 1 ]++;
+		splitVersion[ splitVersion.length - 1 ] ++;
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: I fixed space before `++` and `--` in the follow up commit. I need to fix reformat code in my IDE :)

File: client/state/sites/selectors.js [hunk 6]
```diff
@@ -550,3 +553,311 @@ export function getSitePostsPage( state, siteId ) {
...
+	// for version lower than 3.4, we cannot not determine canManage, we'll assume they can
+	return true;
+}
+
+/**
+ * Determines if a Jetpack site can update its files.
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} if the site can update
```
Comment (gziolo) [inline client/state/sites/selectors.js]: Does it make sense to rename to: `if the site can update *its files*`?

File: client/state/sites/selectors.js [hunk 7]
```diff
@@ -550,3 +553,311 @@ export function getSitePostsPage( state, siteId ) {
...
+	);
+}
+
+/**
+ * Determines if a Jetpack site can auto update its files.
+ * This function checks if the given Jetpack site can update its files and if the automatic updater is enabled
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} if the site can auto update
```
Comment (gziolo) [inline client/state/sites/selectors.js]: Does it make sense to rename to: if the site can auto  update *its files*?

:)

File: client/state/sites/selectors.js [hunk 8]
```diff
@@ -550,3 +553,311 @@ export function getSitePostsPage( state, siteId ) {
...
+
+/**
+ * Returns an explanation on why updates are disabled on a Jetpack Site
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @param {String} action The update action we wanted to perform on this site
+ * @return {?String} The reason why file update is disabled
+ */
+export function getJetpackSiteFileModDisableReasons( state, siteId, action = 'modifyFiles' ) {
```
Comment (gziolo) [inline client/state/sites/selectors.js]: Should we update this name to `getJetpackSiteUpdateFilesDisabledReasons` to align with `canJetpackSiteUpdateFiles`? Any other ideas how to rename `Mod` part?

File: client/state/sites/test/selectors.js [hunk 9]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+						is_multi_network: false,
+						unmapped_url: 'https://example.wordpress.com',
+						main_network_site: 'https://example.wordpress.com',
+						file_mod_disabled: []
+					}
+				}
+			} );
+
+			const canUpdateFiles = canJetpackSiteUpdateFiles( state, siteId );
+			expect( canUpdateFiles ).to.equal( false );
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: Code check asserts `false`. Test description says `true`. Probably it needs some tweaks.

File: client/state/sites/test/selectors.js [hunk 10]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+						jetpack_version: '3.4'
+					}
+				}
+			} );
+
+			const canAutoUpdateFiles = canJetpackSiteAutoUpdateFiles( state, siteId );
+			expect( canAutoUpdateFiles ).to.equal( true );
+		} );
+
+		it( 'it should return `true` if the `file_mod_disabled` option contains `automatic_updater_disabled`', () => {
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: Test description needs update:
`it should return `false` if ...`

File: client/state/sites/test/selectors.js [hunk 11]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+				}
+			} );
+
+			const hasMinimumVersion = siteHasMinimumJetpackVersion( state, siteId );
+			expect( hasMinimumVersion ).to.equal( true );
+		} );
+
+		it( 'it should return `true` if jetpack version is equal to minimum version', () => {
+			const jetpackMinVersion = config( 'jetpack_min_version' );
+			const equalVersion = jetpackMinVersion;
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: Those two lines can be merged:
```js
const equalVersion = config( 'jetpack_min_version' );
```

File: client/state/sites/test/selectors.js [hunk 12]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+		it( 'should return `null` for a non-existing site', () => {
+			const hasCustomDomain = hasJetpackSiteCustomDomain( stateWithNoItems, nonExistingSiteId );
+			expect( hasCustomDomain ).to.equal( null );
+		} );
+
+		it( 'it should return `true` if `URL` and `unmapped_url` have the same domain', () => {
+			const state = createStateWithItems( {
+				[ siteId ]: {
+					ID: siteId,
+					URL: 'https://jetpacksite.me',
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: `unmapped_url` and `URL` have different domains. Most likely test description needs an update ;)

File: client/state/sites/test/selectors.js [hunk 13]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+						unmapped_url: 'https://jetpack.co'
+					}
+				}
+			} );
+
+			const hasCustomDomain = hasJetpackSiteCustomDomain( state, siteId );
+			expect( hasCustomDomain ).to.equal( true );
+		} );
+
+		it( 'it should return `false` if `URL` and `unmapped_url` have different domains', () => {
```
Comment (gziolo) [inline client/state/sites/test/selectors.js]: `unmapped_url` and `URL` use the same domains. Most likely test description needs an update ;)

**Comment by gziolo** [comment]
Oh, I didn't find time to compare old functions and their new implementations.

File: client/state/sites/selectors.js (hunk 8 above)
Comment (Tug) [inline client/state/sites/selectors.js]: Hmm, I'm not the original author of this code and it might have made more sense for them to call it "FileMod" since its closer to the option name. Ok let's switch to `getJetpackSiteUpdateFilesDisabledReasons` for consistency.

File: client/state/sites/test/selectors.js [hunk 14]
```diff
@@ -1184,4 +1228,778 @@ describe( 'selectors', () => {
...
+				}
+			} );
+
+			const canUpdateFiles = canJetpackSiteUpdateFiles( state, siteId );
+			expect( canUpdateFiles ).to.equal( null );
+		} );
+
+		it( 'it should return `false` if jetpack version is smaller than minimum version', () => {
+			const jetpackMinVersion = configStub( 'jetpack_min_version' );
+			const smallerVersion = changeVersion( jetpackMinVersion, -1 );
```
Comment (Tug) [inline client/state/sites/test/selectors.js]: That's @retrofox work ;)

**Comment by Tug** [comment]
@gziolo awesome review, you did spot some important mistakes in those tests.
I hope I fixed them now :)

File: client/state/sites/selectors.js [hunk 15]
```diff
@@ -550,3 +553,311 @@ export function getSitePostsPage( state, siteId ) {
 export function getSiteFrontPageType( state, siteId ) {
 	return getSiteOption( state, siteId, 'show_on_front' );
 }
+
+/**
+ * Determines if a Jetpack site has opted in for full-site management
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} if the site can be managed from calypso
```
Comment (gziolo) [inline client/state/sites/selectors.js]: Yes, makes sense to leave it as it is :)

**Comment by gziolo** [comment]
It looks good from my perspective. It would be nice to have more 👀  on this one. It's huge diff :)

File: client/state/sites/selectors.js [hunk 16]
```diff
@@ -550,3 +553,307 @@ export function getSitePostsPage( state, siteId ) {
...
+			clue === 'automatic_updater_disabled' ) {
+			return i18n.translate( 'Any autoupdates are explicitly disabled by a site administrator.' );
+		}
+
+		if ( action === 'autoupdateCore' &&
+			clue === 'wp_auto_update_core_disabled' ) {
+			return i18n.translate( 'Core autoupdates are explicitly disabled by a site administrator.' );
+		}
+		return null;
+	} ).filter( reason => reason );
```
Comment (ockham) [inline client/state/sites/selectors.js]: Maybe use lodash's `compact` instead of `filter` with identity.

File: client/state/sites/test/selectors.js [hunk 17]
```diff
@@ -1184,4 +1228,791 @@ describe( 'selectors', () => {
...
+			const canManage = canJetpackSiteManage( state, siteId );
+			expect( canManage ).to.equal( false );
+		} );
+	} );
+
+	describe( '#canJetpackSiteUpdateFiles()', () => {
+		let configStub;
+
+		useSandbox( ( sandbox ) => {
+			configStub = sandbox.stub().withArgs( 'jetpack_min_version' ).returns( '3.3' );
```
Comment (ockham) [inline client/state/sites/test/selectors.js]: Can we use `sinon.stub()` instead of a full sandbox?

File: client/state/sites/test/selectors.js (hunk 14 above)
Comment (ockham) [inline client/state/sites/test/selectors.js]: I don't, really. We're explicitly setting `jetpack_min_version` in the config stub anyway, so why not make the "lesser version" explicit, too (instead of adding a somewhat complex function to decrement that version)? I'd much rather remove that additional utility function, TBH.

File: client/state/sites/test/selectors.js [hunk 18]
```diff
@@ -1184,4 +1228,791 @@ describe( 'selectors', () => {
...
+			const canAutoUpdateFiles = canJetpackSiteAutoUpdateFiles( state, siteId );
+			expect( canAutoUpdateFiles ).to.equal( false );
+		} );
+	} );
+
+	describe( '#siteHasMinimumJetpackVersion()', () => {
+		let configStub;
+
+		useSandbox( ( sandbox ) => {
+			configStub = sandbox.stub().withArgs( 'jetpack_min_version' ).returns( '3.3' );
```
Comment (ockham) [inline client/state/sites/test/selectors.js]: Same as above -- maybe use `sinon.stub()` instead of a full sandbox?

File: client/state/sites/test/selectors.js [hunk 19]
```diff
@@ -1184,4 +1228,791 @@ describe( 'selectors', () => {
...
+						unmapped_url: 'https://example.wordpress.com'
+					}
+				}
+			} );
+
+			const isSecondary = isJetpackSiteSecondaryNetworkSite( state, siteId );
+			expect( isSecondary ).to.equal( false );
+		} );
+
+		it( 'it should return `true` for multisite sites which unmapped_url does not match with main_network_site', () => {
```
Comment (ockham) [inline client/state/sites/test/selectors.js]: Maybe reword to `whose unmapped_url does not match their main_network_site`

**Comment by ockham** [comment]
Awesome work! Don't let the amount of my comments get you scared, they're mostly nitpicks. I really appreciate your work, and it's great that everything is covered so well by tests!

File: client/state/sites/selectors.js [hunk 20]
```diff
@@ -550,3 +553,307 @@ export function getSitePostsPage( state, siteId ) {
...
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {?Boolean} if the site can be managed from calypso
+ */
+export function canJetpackSiteManage( state, siteId ) {
+	// for versions 3.4 and higher, canManage can be determined by the state of the Manage Module
+	const siteJetpackVersion = getSiteOption( state, siteId, 'jetpack_version' );
+
+	if ( ! siteJetpackVersion ) {
+		return null;
```
Comment (Tug) [inline client/state/sites/selectors.js]: I'm not sure, I tried to stay as close from the original implementation as possible.
Maybe at some point a jetpack site can be connected but all the information has not been synchronized yet. I haven't had a look into the jetpack components yet but they might just use this behaviour.

Comment (Tug) [inline client/state/sites/selectors.js]: Just had a look and it doesn't seem to be the case...
Fixing 👍

**Comment by Tug** [comment]
@ockham Thanks for reviewing, I fixed most of the issues you raised 👍 .
We still need to decide on how we should handle undefined behaviors: what should a selector return for a non existing site, a non-jetpack site given to a jetpack selector, etc?
Null doesn't sound too bad I find.

File: client/state/sites/selectors.js [hunk 21]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+}
+
+/**
+ * Determines if the Jetpack plugin of a Jetpack Site has menus
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site has Jetpack menus management
+ */
+export function hasJetpackSiteJetpackMenus( state, siteId ) {
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: Should this function return `null` if a non-Jetpack `siteId` is given?

File: client/state/sites/selectors.js [hunk 22]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+	// for version lower than 3.4, we cannot not determine canManage, we'll assume they can
+	return true;
+}
+
+/**
+ * Determines if a Jetpack site can update its files.
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site can update its file
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: We should probably indicate that this function returns `null` as well.

File: client/state/sites/selectors.js [hunk 23]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+	return withoutHttp( unmappedUrl ) !== withoutHttp( mainNetworkSite );
+}
+
+/**
+ * Determines if all given modules are active for a Jetpack Site
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @param {Array} moduleIds A list of active module ids to verify
+ * @return {Boolean} true if the all the given modules are active for this site
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: We should probably note that this function returns null as well.

File: client/state/sites/selectors.js [hunk 24]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+	// Compare unmapped_url with the main_network_site to see if is the main network site.
+	return withoutHttp( unmappedUrl ) === withoutHttp( mainNetworkSite );
+}
+
+/**
+ * Determines if a Jetpack site is a secondary network site
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site is a secondary network site
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: We should probably note that this function returns null as well.

File: client/state/sites/selectors.js [hunk 25]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+}
+
+/**
+ * Determines if a site is the main site in a Network
+ * True if it is either in a non multi-site configuration
+ * or if its url matches the `main_network_site` url option
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site is the main site
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: We should probably note that this function returns null as well.

File: client/state/sites/selectors.js [hunk 26]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+	} ) );
+}
+
+/**
+ * Return true is the given Jetpack site has a version equal or greater than
+ * the minimum Jetpack version as set by the 'jetpack_min_version' config value.
+ *
+ * @param  {Object} state - whole state tree
+ * @param  {Number} siteId - site id
+ * @return {Boolean} true if the site has minimum jetpack version
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: We should probably note that this function returns null as well.

File: client/state/sites/selectors.js [hunk 27]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+
+/**
+ * Determines if a Jetpack site can auto update its files.
+ * This function checks if the given Jetpack site can update its files and if the automatic updater is enabled
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site can auto update
+ */
+export function canJetpackSiteAutoUpdateFiles( state, siteId ) {
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: If we decide to go with `null` returns, this function should probably return `null`.

File: client/state/sites/selectors.js [hunk 28]
```diff
@@ -550,3 +554,304 @@ export function getSitePostsPage( state, siteId ) {
...
+
+/**
+ * Determines if the Jetpack plugin of a Jetpack Site has themes
+ *
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @return {Boolean} true if the site has Jetpack themes management
+ */
+export function hasJetpackSiteJetpackThemes( state, siteId ) {
+	const siteJetpackVersion = getSiteOption( state, siteId, 'jetpack_version' );
```
Comment (drewblaisdell) [inline client/state/sites/selectors.js]: If we decide to go with null returns, this function should probably return `null`.

**Comment by Tug** [comment]
@drewblaisdell thanks for reviewing!

I think I'll stick with returning `null` in those cases.

No selector is currently throwing errors at the moment in calypso and this is not something a developer might expect here.

Moreover, if we look at the current selectors we can see that the default is to return `null` when the input does not match any data in the current state (see [`getRawSite`](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/selectors.js#L41), [`getSite`](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/selectors.js#L58), [`isSingleUserSite`](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/selectors.js#L116), [`isJetpackSite`](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/selectors.js#L130) ...).

That said I agree that we should update the JSDoc comments to make it clear that it can return a boolean or `null`.

**Comment by Tug** [comment]
Updated JSDoc comments and rebased :)
This is ready for another review!

File: client/state/sites/selectors.js [hunk 29]
```diff
@@ -561,3 +565,335 @@ export function getSiteFrontPageType( state, siteId ) {
...
+	return (
+		! includes( fileModDisabled, 'disallow_file_mods' ) &&
+		! includes( fileModDisabled, 'has_no_file_system_write_access' )
+	);
+}
+
+/**
+ * Determines if a Jetpack site can auto update its files.
+ * This function checks if the given Jetpack site can update its files and if the automatic updater is enabled.
+ * Returns null if the site is not known or is not a Jetpack site.
```
Comment (gwwar) [inline client/state/sites/selectors.js]: You can mark in jsdoc a null or x type with:
```
@return {?Boolean}
```

**Comment by lezama** [comment]
Thanks for working on this Tug

cc @johnHackworth @enejb

File: client/state/sites/test/selectors.js [hunk 30]
```diff
@@ -1253,4 +1274,793 @@ describe( 'selectors', () => {
...
+					jetpack: false,
+				}
+			} );
+
+			const canUpdateFiles = canJetpackSiteUpdateFiles( state, siteId );
+			expect( canUpdateFiles ).to.equal( null );
+		} );
+
+		it( 'it should return `false` if jetpack version is smaller than minimum version', () => {
+			const smallerVersion = '3.2';
```
Comment (gwwar) [inline client/state/sites/test/selectors.js]: We should probably stub the config value ` "jetpack_min_version": "3.3",` so this test can make sense in context and won't break when we update the config value

**Comment by gwwar** [comment]
Thanks @Tug and @retrofox. Since none of these selectors are in use let's aim to land this. We can keep a close eye on new usages in smaller future PRs.

**Comment by tyxla** [comment]
Thanks for the awesome work here, @Tug!
